cub/test/catch2_test_device_copy_batched.cu

```diff
@@ -0,0 +1,231 @@
+// SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cub/device/device_copy.cuh>
+#include <cub/util_macro.cuh>
+
+#include <thrust/detail/raw_pointer_cast.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/transform.h>
+
+#include <cstdint>
+
+#include "catch2_large_problem_helper.cuh"
+#include "catch2_test_device_memcpy_batched_common.cuh"
+#include "catch2_test_launch_helper.h"
+#include <c2h/catch2_test_helper.h>
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
+DECLARE_LAUNCH_WRAPPER(cub::DeviceCopy::Batched, copy_batched);
+
+/**
+ * @brief Function object class template that takes an offset and returns an iterator at the given
+ * offset relative to a fixed base iterator.
+ *
+ * @tparam IteratorT The random-access iterator type to be returned
+ */
+template <typename IteratorT>
+struct offset_to_transform_it
+{
+  template <typename OffsetT>
+  __host__ __device__ __forceinline__ auto operator()(OffsetT offset) const
+  {
+    return thrust::make_transform_output_iterator(base_it + offset, ::cuda::std::identity{});
+  }
+  IteratorT base_it;
+};
+
+template <typename T>
+struct offset_to_constant_it
+{
+  template <typename OffsetT>
+  __host__ __device__ __forceinline__ auto operator()(OffsetT offset) const
+  {
+    return thrust::make_constant_iterator(static_cast<T>(offset));
+  }
+};
+
+struct object_with_non_trivial_ctor
+{
+  static constexpr std::int32_t magic_constant = 923390;
+
+  std::int32_t field;
+  std::int32_t magic;
+
+  __host__ __device__ object_with_non_trivial_ctor()
+  {
+    magic = magic_constant;
+    field = 0;
+  }
+  __host__ __device__ object_with_non_trivial_ctor(std::int32_t f)
+  {
+    magic = magic_constant;
+    field = f;
+  }
+
+  object_with_non_trivial_ctor(const object_with_non_trivial_ctor& x) = default;
+
+  __host__ __device__ object_with_non_trivial_ctor& operator=(const object_with_non_trivial_ctor& x)
+  {
+    if (magic == magic_constant)
+    {
+      field = x.field;
+    }
+    return *this;
+  }
+};
+
+static std::ostream& operator<<(std::ostream& os, const object_with_non_trivial_ctor& val)
+{
+  os << '(' << val.field << ',' << val.magic << ')';
+  return os;
+}
```

Comment:
Remark: the prefereed way to define `operator<<` is as hidden friend inside `object_with_non_trivial_ctor`, but not a big issue here. Same for `operator==`.